### PR TITLE
Use the scoped keyword in LibraryImportGenerator

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Use the scoped keyword to significantly simplify the caller-allocated-buffer logic for stateful marshallers while also increasing safety/keeping us from shooting ourselves in the foot with lifetimes.

This PR also updates all of the in-box generators to build against 4.4 and fixes any build errors that were encountered as part of that process.

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ManagedTypeInfo.cs [hunk 1]
```diff
@@ -15,7 +15,17 @@ namespace Microsoft.Interop
...
-        public TypeSyntax Syntax { get; } = SyntaxFactory.ParseTypeName(FullTypeName);
+        private TypeSyntax? _syntax;
+        public TypeSyntax Syntax => _syntax ??= SyntaxFactory.ParseTypeName(FullTypeName);
+
+        protected ManagedTypeInfo(ManagedTypeInfo original)
+        {
+            FullTypeName = original.FullTypeName;
+            DiagnosticFormattedName = original.DiagnosticFormattedName;
+            // Explicitly don't initialize _syntax here. We want Syntax to be recalculated
+            // from the results of a with-expression, which assigns the the new property values
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ManagedTypeInfo.cs:26]:
```suggestion
            // from the results of a with-expression, which assigns the new property values
```

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/StatefulMarshallingStrategy.cs [hunk 2]
```diff
@@ -140,10 +140,17 @@ public IEnumerable<StatementSyntax> GenerateUnmarshalCaptureStatements(TypePosit
...
         {
             // <marshaller> = new();
-            yield return MarshallerHelpers.Declare(
-                _marshallerTypeSyntax,
+            LocalDeclarationStatementSyntax declaration = MarshallerHelpers.Declare(
+                _marshallerType.Syntax,
                 context.GetAdditionalIdentifier(info, MarshallerIdentifier),
                 ImplicitObjectCreationExpression(ArgumentList(), initializer: null));
+
+            if (_marshallerType is ValueTypeInfo { IsByRefLike: true })
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/StatefulMarshallingStrategy.cs:148]: This warrants a comment.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
Closing and reopening to retrigger all legs now that packages have been mirrored

**Comment by elinor-fung** [comment]
Would it make sense to separate out the version upgrade from actually using the scoped keyword in LibraryImportGenerator at this point? It seems like (fallout from) the version upgrade has gotten more involved and become the bulk of this change now, so it might be nice to keep them separate.

**Comment by jkoritzinsky** [comment]
I was thinking about that... I'll spin it out to another PR once I get CI green (so I don't need to iterate on it much).

**Comment by jkoritzinsky** [comment]
Moved the code-style changes to https://github.com/dotnet/runtime/pull/74445. @elinor-fung can I get another quick review look?

**Review by elinor-fung — APPROVED** [review]

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs [hunk 3]
```diff
@@ -28,10 +28,7 @@ internal sealed partial class TarHeader
             archiveStream.ReadExactly(buffer);
 
             TarHeader? header = TryReadAttributes(initialFormat, buffer);
-            if (header != null)
-            {
-                header.ProcessDataBlock(archiveStream, copyData);
-            }
+            header?.ProcessDataBlock(archiveStream, copyData);
```
Comment (elinor-fung) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:31]: This should be part of the code style changes, right?

Comment (jkoritzinsky) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:31]: Yes, it should. I'll revert this change.

**Comment by jkoritzinsky** [comment]
Failure is slow macs and https://github.com/dotnet/runtime/issues/73688. Merging

**Pull request merged**